### DISCUSSION TITLE
Do you use react native storybook server? Whats your usecase?

### DISCUSSION THREAD

**Discussion by dannyhw** [storybookjs/react-native#207] [General]
I'm considering replacing react native storybook server with a different solution and I'm curious what current users are using it for and how I could include that in the new solution.

Currently the server causes a lot of confusion and it can be quite buggy. It also is completely coupled with the ondevice UI and can't be used on its own.

I want to create a react-native-web storybook with all the web storybook addons and some of the features from the server. It would work on its own by rendering react-native-web previews where possible and could be deployed on the web. 

Then perhaps via addons you could also have control over ondevice storybook similar to the server currently. 

Currently this is in the ideas phase and nothing has been implemented yet. Please give your opinions on this, all feedback is welcome.

**Comment by sregg** [comment]
I used to use the server to select stories using a VSCode extension. In theory, it was supposed to be better than searching and selecting stories from the in-app UI but in practice, starting the server was always a pain.
I'd love to have a web version of my React-Native Storybook.
React-Native-Web should work 90% of the time but the remaining 10% (native dependency, weird css mismatch, etc...) will make our lives miserable.
We're gonna try with our design system library first. I'll report here once we have completed that spike.

**Reply by dannyhw** [reply]
Hey, thanks for providing this feedback. I hadn't heard of this vscode extension actually, seems interesting. To some degree I think that this could actually be improved by just having a better ondevice UI (making it easier to find the story). The react-native-web version of storybook I think would be specifically nice for using reactjs addons and tools built for the web. However perhaps we can bring back this vscode extension if the server functionality ends up working well in the react-native-web version. 

To some degree you can already run reactjs storybook on react native stories if you manually configure it however I think there is potential to add official support for this and react native specific addons etc.

**Comment by pvinis** [comment]
I'm using the server to control what the stories show on the device. I find the ondevice UI very slow and awkward to use. changing knobs and looking at actions is waaaaaaaay easier and faster with the server.

**Reply by dannyhw** [reply]
Thanks for your feedback!

How could the ondevice UI be improved to make it more convenient? If it was improved would you still use the server?

Do you use storybook during development or in another way?

**Reply by pvinis** [reply]
I use storybook for development only. Well, I do have it linked to show off some stuff, but it's mainly for development/documentation by me and the rest of the team.

I don't know if ondevice UI could be good enough, to be honest.
- I don't like the fact that it takes space.
The mobile screen is small, I don't want any less than what I can get. Also, when messing with tabs, or navbars etc, it's just a pain to have to make changes in my stories in order to avoid the ondevice UI.
- I don't like that fact that it loses the page I'm on (I guess that could be fixed, but server works well here).
- I don't like that I can't see all the things at once.
This ties in with the first a bit. I like the server because I can see the list of stories, I can see the knobs, the actions, everything, in one place, all at once, and with the simulator next to the browser, I see the storybook too. With ondevice UI, I have to choose one of the above at any one time. When I develop and want to mess around with knobs or see actions etc, I don't want to click 4 times on the simulator to see what's up. I want to click 0. And the server gives me that.

I would be open to any solutions for these, but it feels like the ondevice UI is kinda doomed for me 🤔.

**Reply by dannyhw** [reply]
Thanks for providing more details on your setup this will help with deciding on future work regarding the server 👍 .

One thing I want to say though is that the ondeviceUI is useful for other use cases and won't be going away. The ondevice UI currently is the only way to distribute the react native storybook to testers, designers etc. That's because it's not possible to deploy the server in a distributed way and that's just not easily solvable because of the way it's setup. With the ondeviceUI this is easy to solve by just distributing the ondevice version as an app via expo or firebase etc. 

That being said, I don't disagree with your points and I can see how for a development workflow having a way to externally control the storybook is useful. My intention is to find a way to better support both approaches in a way that is more maintainable and not tightly coupled. What I mean by that is I want the server to function without the ondeviceUI and the other way around.

Just a disclaimer I still don't really know the server code at all so bringing the server or a replacement to 6.0 is likely going to take a while. The @storybook/react-native 6.0 might be out first before the server returns depending on how development goes.

**Reply by pvinis** [reply]
If you have ideas/plans, I'm willing to give some hours here and there to help.

**Comment by williamgoulois** [comment]
Hi ! I'm using storybook server during development as a TDD visual tool and i would love to be able to change stories on both emulators/devices at the same time with web interface.
It would be great to be able to change controls from the server too.

Right now it's pretty painful i have to change everything two times : 
- stories on my android device and my ios emulator
- controls on both devices

**Reply by dannyhw** [reply]
Thanks for providing feedback! I'm getting the impression now that the server needs to be part of the project, however I do think we can do it in a different way. 
I'll consider this use-case when experimenting with potential implementations 👍
